src/main/java/com/yelp/nrtsearch/server/luceneserver/script/RuntimeScript.java

```diff
@@ -0,0 +1,185 @@
+    return segmentDocLookup;
+  }
+
+  /**
+   * Simple abstract implementation of a {@link ValueSource} this can be extended for engines that
+   * need to implement a custom {@link RuntimeScript}. The newInstance and needs_score must be
+   * implemented. If more state is needed, the equals/hashCode should be redefined appropriately.
+   *
+   * <p>This class conforms with the script compile contract, see {@link ScriptContext}. However,
+   * Engines are also free to create there own {@link ValueSource} implementations instead.
+   */
+  public abstract static class SegmentFactory extends ValueSource {
+    private final Map<String, Object> params;
+    private final DocLookup docLookup;
+
+    public SegmentFactory(Map<String, Object> params, DocLookup docLookup) {
+      this.params = params;
+      this.docLookup = docLookup;
+    }
+
+    public Map<String, Object> getParams() {
+      return params;
+    }
+
+    public DocLookup getDocLookup() {
+      return docLookup;
+    }
+
+    /**
+     * Create a {@link FunctionValues} instance for the given lucene segment.
+     *
+     * @param ctx Context map for
+     * @param context segment context
+     * @return script to produce values for the given segment
+     */
+    public abstract FunctionValues newInstance(Map ctx, LeafReaderContext context);
+
+    /**
+     * Get if this script will need access to the document score.
+     *
+     * @return if this script uses the document score.
+     */
+    public abstract boolean needs_score();
+
+    /** Redirect {@link ValueSource} interface to script contract method. */
+    @Override
+    public FunctionValues getValues(Map context, LeafReaderContext ctx) throws IOException {
+      return newInstance(context, ctx);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(params, docLookup);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj == null) {
+        return false;
+      }
+      if (obj.getClass() != this.getClass()) {
+        return false;
+      }
+      RuntimeScript.SegmentFactory factory = (RuntimeScript.SegmentFactory) obj;
+      return Objects.equals(factory.params, this.params)
+          && Objects.equals(factory.docLookup, this.docLookup);
+    }
+
+    @Override
+    public String toString() {
+      return "RuntimeScriptValuesSource: params: " + params + ", docLookup: " + docLookup;
+    }
+  }
+
+  /**
+   * Factory required from the compilation of a ScoreScript. Used to produce request level {@link
+   * ValueSource}. See script compile contract {@link ScriptContext}.
+   */
+  public interface Factory {
+    /**
+     * Create request level {@link SegmentFactory}.
+     *
+     * @param params parameters from script request
+     * @param docLookup index level doc value lookup provider
+     * @return {@link ValueSource} to evaluate script
+     */
+    ValueSource newFactory(Map<String, Object> params, DocLookup docLookup);
```

Comment:
I'm not sure you are really getting much using `ValueSource`. `SegmentFactory` could be a simple interface.

The `RuntimeFieldDef` could just provide a `RuntimeScript.SegmentFactory` implementation.

Reply:
Thanks. Converting to SegmentFactory was much simpler and more straightforward.